### PR TITLE
[v2] break out pages.json and json data paths from app bundle

### PR DESCRIPTION
**Pull request by pieh**
inline needed data for page in html and delay loading rest of it after initial mount has been completed

### CODE REVIEW

File: packages/gatsby/cache-dir/loader.js [hunk 1]
```diff
@@ -9,14 +9,27 @@ let inInitialRender = true
...
 let pathPrefix = ``
 let fetchHistory = []
+let fetchingAllPagesPromise = null
 const failedPaths = {}
 const failedResources = {}
 const MAX_HISTORY = 5
 
 const jsonStore = {}
 
+const fetchAllPagesMap = () => {
```
Comment (KyleAMathews) [inline packages/gatsby/cache-dir/loader.js]: maybe `fetchPageResourceMap` would be clearer name?

**Comment by pieh** [comment]
Yes, this is how it works with this.
(highlighted resource is `data.json` which contains content of `pages.json` and paths to json data):
network-waterfall
Btw, seems like we need to take a look at preloading css after webpack 4 merge

I need to make one more change (not counting renaming and adding some comments) which I realized few hours ago: need to handle case when someone clicks a link before `data.json` is loaded.

**Comment by KyleAMathews** [comment]
Nice!

This PR will take care of the double loading of the inlined CSS https://github.com/webpack-contrib/mini-css-extract-plugin/pull/55

**Comment by KyleAMathews** [comment]
This issue is tracking the remaining work for mini-css-extract-plugin https://github.com/gatsbyjs/gatsby/issues/4677

**Comment by robwierzbowski** [comment]
> Btw, seems like we need to take a look at preloading css after webpack 4 merge

Does that mean Gatsby 2.0 will be on Webpack 4?

**Comment by pieh** [comment]
Rebased and resolved conflict. It is feature complete, so after review could be merged.

File: packages/gatsby/cache-dir/loader.js [hunk 2]
```diff
@@ -167,12 +187,28 @@ const queue = {
   addProdRequires: prodRequires => {
     asyncRequires = prodRequires
   },
-
+  addDataPaths: dataPaths => {
+    jsonDataPaths = dataPaths
+  },
   dequeue: () => resourcesArray.pop(),
-  enqueue: rawPath => {
+  enqueue: async rawPath => {
```
Comment (KyleAMathews) [inline packages/gatsby/cache-dir/loader.js]: we haven't been using async/await as the polyfills are super heavy

Comment (KyleAMathews) [reply packages/gatsby/cache-dir/loader.js]: Which is tragic 😢

Comment (pieh) [reply packages/gatsby/cache-dir/loader.js]: Right, callbacks it is then!

**Comment by pieh** [comment]
Rebased and changed `async/await` to `then`. Should be ready for another review

**Comment by KyleAMathews** [comment]
Nice!

**Pull request merged**